### PR TITLE
Feature pictures multi dive

### PR DESCRIPTION
**Pull request by sfuchs79**
### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

In the dive picture tab show pictures of all selected dive.
But at the same moment take care that in the profile only
pictures from displayed_dive are displayed.

With @bstoeger help and his cleanup work (which is already in master) I was able to finalize this.

Plus I added a few fine tuning items (please see commits!).

I have to admit that I didn't do that much testing up to now. So someone else please also has to test this a little bit.

### Related issues:

#898 

### Release note:

Release notes to be updated.

### Mentions:

@dirkhh @bstoeger @atdotde

### CODE REVIEW

**Comment by dirkhh** [comment]
That looks deceptively simple and seems to integrate no new strings...
I'll test it later today and decide what to do :-)

**Comment by sfuchs79** [comment]
@dirkhh 
**But this really doesn't work yet - see my comments!**
I put it here as an PR because it's **maybe** not too far away from being functional. And maybe someone more skilled than me only needs minutes to finalize it. But it **could be** also the case that a lot more changes are needed to really finalize it.
But I'm a big fan of the idea that you once build it and test it so you can better understand whats currently not working yet.
BTW: Is there a good online tutorial/docu/whatever where I could learn to understand Qt/C++ a little bit better? Focus should be mainly on Qt for e.g. understanding all this widgets, models, signals, slots, whatever stuff.

**Comment by bstoeger** [comment]
The crash wasn't your fault, see #911 
Still, deletion doesn't work, because `dive_remove_picture` only accesses the current dive. You'd either have to remember the dive the picture belongs to or loop over all (selected) dives.

File: qt-models/divepicturemodel.cpp [hunk 1]
```diff
@@ -56,25 +56,34 @@ void DivePictureModel::updateDivePictures()
...
+			numberOfPicturesOld = numberOfPictures;
+			numberOfPictures += dive_get_picture_count(dive);
+			if (dive->id == displayed_dive.id)
+				rowDDEnd = numberOfPictures;
+			FOR_EACH_PICTURE(dive) {
+				stringPixmapCache[QString(picture->filename)].offsetSeconds = picture->offset.seconds;
+				pictures.push_back(picture);
+			}
+		}
+		QList<SPixmap> list = QtConcurrent::blockingMapped(pictures, scaleImages);
```
Comment (bstoeger) [inline qt-models/divepicturemodel.cpp]: Note that you are repeating this for *every* dive. I think this code should go up one level...

**Comment by bstoeger** [comment]
To make all pictures deletable, you might try something like this (includes fix of #911):
```
diff --git a/core/dive.c b/core/dive.c
index 033139d0c..fc40d746d 100644
--- a/core/dive.c
+++ b/core/dive.c
@@ -3838,17 +3838,20 @@ struct picture *clone_picture(struct picture *src)
        return dst;
 }
 
-void dive_remove_picture(char *filename)
+// Remove true if picture was found and deleted
+bool dive_remove_picture(struct dive *d, char *filename)
 {
-       struct picture **picture = &current_dive->picture_list;
-       while (picture && !same_string((*picture)->filename, filename))
+       struct picture **picture = &d->picture_list;
+       while (*picture && !same_string((*picture)->filename, filename))
                picture = &(*picture)->next;
-       if (picture) {
+       if (*picture) {
                struct picture *temp = (*picture)->next;
                picture_free(*picture);
                *picture = temp;
                invalidate_dive_cache(current_dive);
+               return true;
        }
+       return false;
 }
 
 /* this always acts on the current divecomputer of the current dive */
diff --git a/core/dive.h b/core/dive.h
index 2b227690b..fa776846f 100644
--- a/core/dive.h
+++ b/core/dive.h
@@ -421,7 +421,7 @@ extern struct picture *clone_picture(struct picture *src);
 extern bool dive_check_picture_time(struct dive *d, int shift_time, timestamp_t timestamp);
 extern void dive_create_picture(struct dive *d, const char *filename, int shift_time, bool match_all);
 extern void dive_add_picture(struct dive *d, struct picture *newpic);
-extern void dive_remove_picture(char *filename);
+extern bool dive_remove_picture(struct dive *d, char *filename);
 extern unsigned int dive_get_picture_count(struct dive *d);
 extern bool picture_check_valid(const char *filename, int shift_time);
 extern void picture_load_exif_data(struct picture *p);
diff --git a/qt-models/divepicturemodel.cpp b/qt-models/divepicturemodel.cpp
index 0f3054b7b..f4d6c2d8f 100644
--- a/qt-models/divepicturemodel.cpp
+++ b/qt-models/divepicturemodel.cpp
@@ -127,7 +127,11 @@ QVariant DivePictureModel::data(const QModelIndex &index, int role) const
 
 void DivePictureModel::removePicture(const QString &fileUrl, bool last)
 {
-       dive_remove_picture(fileUrl.toUtf8().data());
+       int i;
+       struct dive *dive;
+       for_each_dive (i, dive)
+               if (dive->selected && dive_remove_picture(dive, fileUrl.toUtf8().data()))
+                       break;
        if (last) {
                copy_dive(current_dive, &displayed_dive);
                updateDivePictures();
```

File: qt-models/divepicturemodel.cpp (hunk 1 above)
Comment (sfuchs79) [reply qt-models/divepicturemodel.cpp]: Yes, that's wrong. Thanks for pointing it out.

**Comment by sfuchs79** [comment]
@bstoeger 
I locally fixed the one mistake you pointed out and I will also try the "deleting dives" thing. BTW: You know that you can do on Github a PR on the branch I use for the PR? This makes it easy to combine different work together.

Unfortunately the main issue with this change is still not solved:
If I have two dives A and B with both pictures and if I then select the two dives I would expect to see the following in the picture tab:
Picture dive A # 1
...
Picture dive A # n
Picture dive B # 1
...
Picture dive B # n

Instead what I see are the dive A and dive B pictures mixed together. And this also breaks displaying the right pictures in the profile. And I still don't know why the pictures get "sorted" somehow.

Can you confirm this?

**Comment by bstoeger** [comment]
Well, the fundamental problem is that the ordering of the pictures is based on the hashed filenames - thus by definition is pretty much random.

I tried to address this in #926.

**Comment by sfuchs79** [comment]
With @bstoeger help I was able to push a major update for this.
This version should now really work ;-)

File: qt-models/divepicturemodel.cpp [hunk 2]
```diff
@@ -56,17 +56,29 @@ void DivePictureModel::updateDivePictures()
...
 
-	QtConcurrent::blockingMap(pictures, scaleImages);
+			if (dive->id == displayed_dive.id) {
+				rowDDStart = pictures_count_old;
+				rowDDEnd = pictures.count() - 1;
+			}
 
-	beginInsertRows(QModelIndex(), 0, pictures.count() - 1);
-	endInsertRows();
+			QtConcurrent::blockingMap(pictures, scaleImages);
```
Comment (bstoeger) [inline qt-models/divepicturemodel.cpp]: Please move this up two levels. The point is to call this once with the finished list.

File: qt-models/divepicturemodel.cpp [hunk 3]
```diff
@@ -56,17 +56,29 @@ void DivePictureModel::updateDivePictures()
...
+
+			beginInsertRows(QModelIndex(), pictures_count_old, pictures.count() - 1);
```
Comment (bstoeger) [inline qt-models/divepicturemodel.cpp]: These two lines should also be moved out of the for_each_dive loop.
`pictures_count_old` is not necessary as far as I can tell. The idea of `beginInsertRows()` and `endInsertRows()` is to call them only once per batch-manipulation.

**Comment by sfuchs79** [comment]
@bstoeger Thanks for the review! Updates done.

**Comment by dirkhh** [comment]
Not quite as simple anymore :-)
In order to test this I'd have to import a lot more pictures into my dives. This is one of those things I keep wanting to do anyway, so maybe I should just take the time now :-)
In the meantime - @sfuchs79 and @bstoeger you both have tested this against real data? This is good to go?

**Comment by dirkhh** [comment]
OK, tested and it seems to work fine.
merging

**Comment by bstoeger** [comment]
For more robustness, perhaps we should add a `rowDDStart = rowDDEnd = 0;` to the `if (dive_table.nr == 0)` block? Sure, if no dive is selected, the profile widget will not attempt to draw pictures. Nevertheless, I'd prefer the internal state to be consistent.

**Comment by sfuchs79** [comment]
@dirkhh Thanks for testing!
@bstoeger Yes, you're right. Any call of DivePictureModel::updateDivePictures which is not adding any pictures should better set rowDD* to 0. But isn't it that we even better put it into the first `if`:
https://github.com/sfuchs79/subsurface/blob/bc6adabfbbcf2226d27547cf4e56cd5d7ffe4e0b/qt-models/divepicturemodel.cpp#L53

**Comment by bstoeger** [comment]
That's correct. This would also catch the case where displayed_dive is not in the selected list (for whatever weird reason).

**Pull request merged**